Add a Gradle 4.x upgrade chapter

pledbrook opened the pull request:
This is a provisional chapter that will hopefully evolve. It covers everything I thought was appropriate from the various release notes.

A few question marks remain:

 * Whether to mention the dependency lock scheme, since there is still no migration guide from the Nebula plugin
 * The "From <version>" headings don't seem useful. I'm leaning towards just having headings containing the version number that the corresponding changes were introduced in. So instead of "From 4.3", we would have "4.4".

The nature of some of the content means that the guide needs reviewing and potentially updating every time a new version of Gradle is released until 5.0 comes out.

eriwen commented on subprojects/docs/src/docs/stylesheets/userGuideHtml.xsl:
```diff
@@ -76,7 +76,12 @@
                             <li><a href="https://guides.gradle.org/creating-new-gradle-builds/">Creating New Gradle Builds</a></li>
                             <li><a href="https://guides.gradle.org/creating-build-scans/">Creating Build Scans</a></li>
-                            <li><a href="https://guides.gradle.org/migrating-from-maven/">Migrating From Maven</a></li>
+                            <li><a class="nav-dropdown" data-toggle="collapse" href="#migrating-from" aria-expanded="false" aria-controls="migrating-from">Migrating from...</a>
```
I'm not sure if migration is the right term here. "Upgrade" feels more like it.

pledbrook replied:
Agreed. I was planning to update the nav for the upgrade guide. Now that the new layout has been merged in, I think I'll work against that.

eriwen commented:
@pledbrook I hope you don't mind and our work doesn't conflict, but I went ahead and rebased this atop all the latest docs changes and fixed everything up here: https://github.com/gradle/gradle/pull/6165 — should we continue with that PR?

pledbrook commented:
I have a bunch of changes locally — including fixes for links to fit with the recent doc changes — that I haven't pushed yet because I haven't quite finished with them. I'll work on those today and then compare the result to #6165.

pledbrook commented on subprojects/docs/src/docs/userguide/plugins.adoc:
```diff
@@ -184,6 +184,26 @@ Hello!
 BUILD SUCCEEDED
 ----
 
+[[sec:buildsrc_plugins_dsl]]
```
Yes. The upgrade guide talks about changes to the `plugins {}` block, but I realised that some of those changes weren't actually documented. Hence I added them with the upgrade guide.

eriwen replied:
I'd like to pull these changes out and merge them separately for 4.10.

eriwen commented:
Bah, sorry for the duplicate work then. I used your `api:` link transformer, but ended up with about 9 broken links to fix. Of course there were other non-API links to fix and samples etc.

eriwen commented on subprojects/docs/src/docs/userguide/upgrade_version_4.adoc:
```diff
@@ -0,0 +1,493 @@
+A few pointers about upgrading:
+
+ * Run `gradle help` with the version of Gradle you're upgrading to in order to see any deprecations that apply to your build.
+ * Make sure you update the versions of plugins you use in your build when upgrading the Gradle version. Some plugins will break with newer versions of Gradle.
```
> Some plugins will break with newer versions of Gradle.

Maybe we should specify that some plugins use internal APIs that change between minor versions?

pledbrook replied:
Could do. I'm not sure it's particularly useful information for the user though.

eriwen commented on subprojects/docs/src/docs/userguide/upgrade_version_4.adoc:
```diff
@@ -0,0 +1,493 @@
+ * Make sure you update the versions of plugins you use in your build when upgrading the Gradle version. Some plugins will break with newer versions of Gradle.
+ * In most cases, deprecated features, properties and methods will be removed in Gradle 5.0, resulting in build errors if you try to use them.
+
+The most significant changes include the following:
```
I like this concept of "significant changes" at the top. I wonder if this summary belongs in a more prominent place and make it very visual, such as on gradle.org. 

* The improved POM support will (probably) be enabled by default in Gradle 5.0, thus users may want to upgrade to Gradle 4.10, enable it, then take away that "feature preview" line after upgrading to 5.0.
* I wonder if it would be worthwhile to group "stuff that might break" a little separately from "stuff you should adopt".

pledbrook replied:
I can see value in putting some of the significant changes in marketing material, including on gradle.org. I think it should still remain in some form in this chapter.

I don't have an opinion on the approach users should take with the improved POM support. I guess it allows them to isolate one potential source of breaking changes, but does that gain them much over just upgrading to 5.0 and dealing with the issues one by one?

I agree that keeping new features and breaking changes distinct would be helpful. I'll look at that.

eriwen commented on subprojects/docs/src/docs/userguide/upgrade_version_4.adoc:
```diff
@@ -0,0 +1,493 @@
+Follow the API links to learn how to deal with these deprecations (if no extra information is provided here):
+
+ * link:{javadocPath}/org/gradle/api/tasks/scala/ScalaDocOptions.html#getStyleSheet--[ScalaDocOptions.styleSheet] — the Scaladoc Ant task in Scala 2.11.8 and later no longer support this property.
+ * link:{javadocPath}/org/gradle/api/Nullable.html[Nullable]
```
Perhaps we can suggest an alternative here? The javadoc doesn't say anything.

pledbrook replied:
I don't know if there is an alternative. I'll ask.

eriwen commented on subprojects/docs/src/docs/userguide/upgrade_version_4.adoc:
```diff
@@ -0,0 +1,493 @@
+
+ * link:{javadocPath}/org/gradle/api/tasks/scala/ScalaDocOptions.html#getStyleSheet--[ScalaDocOptions.styleSheet] — the Scaladoc Ant task in Scala 2.11.8 and later no longer support this property.
+ * link:{javadocPath}/org/gradle/api/Nullable.html[Nullable]
+ * link:{javadocPath}/org/gradle/api/Task.html#dependsOnTaskDidWork--[Task.dependsOnTaskDidWork()]
```
Should we say that there's no substitute?

pledbrook replied:
We should say something, yes. I'll review the release notes again and ask others if necessary.

pledbrook replied:
Actually, isn't the Javadoc sufficient here? It explains that users should use declared inputs and outputs instead.

eriwen commented on subprojects/docs/src/docs/userguide/upgrade_version_4.adoc:
```diff
@@ -0,0 +1,493 @@
+ * The <<custom_tasks.adoc#worker_api,Worker API>> for enabling units of work to execute in parallel.
+ * A new, <<feature_lifecycle#sec:incubating_state,incubating>> API for <<#rel4.8:lazy_task_creation,creating and configuring tasks lazily>> that can significantly improve your build's configuration time.
+
+There are many other changes that we list by Gradle version in the rest of the guide.
```
We have not yet covered a few major deprecations that will be removed in Gradle 5.0. We may need a special section, here. 

 * Running Gradle with JDK 7 is deprecated (though tooling API still allows it)
 * `<<` syntax will be removed. Users should use `doLast{}`

I'm sure there are a few other things that we deprecated in 4.0 that may not be fully covered.

pledbrook replied:
OK. I'll ask around. I'll also start updating this page in light of it going live in the 5.0 docs.

pledbrook replied:
I'll try to find out what's happening on this front.

pledbrook commented:
Interesting that you encountered 9 broken API links using the same script. They all worked for me. Wonder what the difference was. Perhaps I'd already fixed the link to begin with. Do you know what was broken about them? Just wondering if something needs fixing in the script.

oehme commented on subprojects/docs/src/docs/userguide/upgrade_version_4.adoc:
```diff
@@ -0,0 +1,493 @@
+
+ * The output produced by a given task is now grouped together, even when other tasks execute in parallel with it.
+ * Task execution headers are printed with a "> Task" prefix.
+ * All output produced during build execution is written to the standard output file handle. This includes messages written to System.err unless you're are redirecting standard error to a file or any other non-console destination.
```
`you're are`

oehme commented on subprojects/docs/src/docs/userguide/upgrade_version_4.adoc:
```diff
@@ -0,0 +1,493 @@
+----
+dependencies {
+    // import a BOM
+    implementation 'org.springframework.boot:spring-boot-dependencies:1.5.8.RELEASE'
```
I think the @gradle/dependency-management wanted to change how this works for 5.0, so they'll want to be aware of this document.

pledbrook replied:
I think most teams will need to review this before it goes live. BTW, if that team does pop in here, the review should be happening at #6165. Sorry for the confusion.

oehme commented on subprojects/docs/src/docs/userguide/upgrade_version_4.adoc:
```diff
@@ -0,0 +1,493 @@
+ * The dependency module is ignored if it is only ever declared as optional.
+ * If the dependency module is also declared elsewhere as not optional, then the constraint derived from the optional dependency declaration is considered when picking the version.
+
+In other words, if an optional dependency has a declared version higher than another, non-optional one, the optional dependency's version is used. However, Gradle does not resolve optional dependencies.
```
> However, Gradle does not resolve optional dependencies.

I think that's the wrong terminology. It does resolve them, that's why version selection now works. What you're trying to say is probably: It doesn't make them part of the graph unless someone else has an actual dependency on them. But that's already mentioned in the paragraph above.

pledbrook replied:
Yes, but I'm trying to clarify the previous paragraph into terminology I find easier to understand. I'll have another go.

pledbrook commented on subprojects/docs/src/docs/userguide/upgrade_version_4.adoc:
```diff
@@ -0,0 +1,493 @@
+
+A few pointers about upgrading:
+
+ * Run `gradle help` with the version of Gradle you're upgrading to in order to see any deprecations that apply to your build.
```
The advantage of `help` is that you don't get a ton of other output that hides the deprecation warnings, but I take your point. I'll incorporate the additional information. I think it's probably worth running `gradle help` first and then move onto `gradle build` (if the task exists). That said, I'm not familiar with how the deprecation warnings appear in a full build. Are they aggregated and displayed together at the beginning/end of the build? Or are they interspersed with the other build output as they are encountered?

oehme replied:
They are shown when they occur. That's no different from "help".

pledbrook replied:
Thanks. As I said, the advantage of `help` is that it doesn't have a lot of output that obfuscates the deprecation warnings, although to be fair, Gradle's normal output isn't particularly verbose either.

Would a `gradle -q --warning-mode=all build` make sense? Or does the `-q` flag not do anything particularly useful in this instance?

oehme replied:
I think you'll want `--warn` instead of `--quiet`. The latter would also silence the warnings.

pledbrook commented on subprojects/docs/src/docs/userguide/upgrade_version_4.adoc:
```diff
@@ -0,0 +1,493 @@
+[[rel4.7:switch_to_publishing_plugins]]
+=== [4.7] Switch to the Maven Publish and Ivy Publish Plugins
+
+Now that the publishing plugins are stable, the <<artifact_management.adoc#artifact_management,legacy publishing>> mechanism based on `Upload` tasks is deprecated. You should replace `upload<Conf>` configuration with a `publishing` block instead. See the <<publishing_overview.adoc#publishing_overview,publishing overview chapter>> for information on how to use the publishing plugins.
```
Why is it not deprecated? I hate to see the perpetuation of two publishing mechanisms.

oehme replied:
Because you can't define your own `Components` and so ecosystems like Android can't use the new publishing mechanism yet. They still rely on `Upload` until we provide that feature.

pledbrook replied:
Ah, yes, I remember about the custom components thing now. Thanks for the reminder. I hope that's implemented in the not too distant future!

oehme commented on subprojects/docs/src/docs/userguide/upgrade_version_4.adoc:
```diff
@@ -0,0 +1,493 @@
+
+=== Potential breaking changes
+
+ * `whenMerged {}` and `beforeMerged {}` blocks for link:{groovyDslPath}/org.gradle.plugins.ide.eclipse.model.EclipseProject.html[EclipseProject] are now executed by Buildship.
```
It makes no sense that they would have this block if they didn't want it to be executed. These users cannot import at the moment for that reason.

pledbrook replied:
Isn't it feasible that a block might have been left over from when `gradle eclipse` was used before users switched to the Buildship import? The import ignored that block, so everything worked and folks forgot about it. Now the import will break with the new version of Gradle.

Is that scenario outside the realm of possibility? It may not be common, but if it's sufficiently likely, I would have thought that justifies putting it down as a potential breaking change. Sometimes fixing broken behaviour _is_ a breaking change.

oehme replied:
> Isn't it feasible that a block might have been left over from when gradle eclipse was used before users switched to the Buildship import?

Those were the users who were blocked because Buildship isn't executing that code and so their project isn't configured correctly.

There is no reasonable scenario where someone wrote that block and didn't want it to be executed. It just makes no sense.

pledbrook replied:
OK, I'll take your word for it. It makes complete sense. I've just had too many experiences where users have setups that don't fit expectations and get burned.

pledbrook replied:
BTW, I don't know if it's possible to remove it from the release notes. @eriwen, any ideas on that front, since it's in a past release notes page rather than the current one?

pledbrook commented:
@oehme I've addressed your feedback in [#6165](https://github.com/gradle/gradle/pull/6165). Please follow up there.

@eriwen I'm closing this pull request to avoid confusion.
